service/handlers_app_user.go

```diff
@@ -0,0 +1,166 @@
+
+// UserIndexHandler is the front page of the web application
+func UserIndexHandler(w http.ResponseWriter, r *http.Request) {
+	page := Page{Title: Environ.Config.Title, Logo: Environ.Config.Logo}
+
+	path := []string{Environ.Config.DocRoot, userIndexTemplate}
+	t, err := template.ParseFiles(strings.Join(path, ""))
+	if err != nil {
+		log.Printf("Error loading the application template: %v\n", err)
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	err = t.Execute(w, page)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+	}
+}
+
+// SystemUserAssertionHandler is the API method to generate a signed system-user assertion for a device
+func SystemUserAssertionHandler(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set("Content-Type", "application/json; charset=UTF-8")
+
+	// Decode the body
+	user := SystemUserRequest{}
+	err := json.NewDecoder(r.Body).Decode(&user)
+	switch {
+	// Check we have some data
+	case err == io.EOF:
+		w.WriteHeader(http.StatusBadRequest)
+		formatBooleanResponse(false, "error-user-data", "", "No system-user data supplied", w)
+		return
+		// Check for parsing errors
+	case err != nil:
+		w.WriteHeader(http.StatusBadRequest)
+		formatBooleanResponse(false, "error-decode-json", "", err.Error(), w)
+		return
+	}
+
+	// Get the model
+	model, err := Environ.DB.GetModel(user.ModelID)
+	if err != nil {
+		logMessage("USER", "invalid-model", "Cannot find model with the selected ID")
+		formatBooleanResponse(false, "invalid-model", "", "Cannot find model with the selected ID", w)
+		return
+	}
+
+	// Check that the model has an active keypair
+	if !model.KeyActive {
+		logMessage("USER", "invalid-model", "The model is linked with an inactive signing-key")
+		formatBooleanResponse(false, "invalid-model", "", "The model is linked with an inactive signing-key", w)
+		return
+	}
+
+	// Create the system-user assertion headers from the request
+	assertionHeaders := userRequestToAssertion(user, model)
+
+	// Sign the system-user assertion
+	signedAssertion, err := Environ.KeypairDB.SignAssertion(asserts.SystemUserType, assertionHeaders, nil, model.AuthorityID, model.KeyID, model.SealedKey)
+	if err != nil {
+		logMessage("USER", "signing-assertion", err.Error())
+		formatBooleanResponse(false, "signing-assertion", "", err.Error(), w)
+		return
+	}
+
+	// Return the signed assertion
+	serializedAssertion := asserts.Encode(signedAssertion)
+
+	response := SystemUserResponse{Success: true, Assertion: string(serializedAssertion)}
+	if err := json.NewEncoder(w).Encode(response); err != nil {
+		logMessage("USER", "signing-assertion", err.Error())
+	}
+}
+
+func userRequestToAssertion(user SystemUserRequest, model Model) map[string]interface{} {
+
+	// Create the salt from the keystore secret
+	reg, _ := regexp.Compile("[^A-Za-z0-9]+")
+	baseSalt := reg.ReplaceAllString(Environ.Config.KeyStoreSecret, "")
+
+	// Encrypt the password
+	salt := fmt.Sprintf("$6$%s$", baseSalt)
```

Comment:
Check the crypt(3) man page: http://man7.org/linux/man-pages/man3/crypt.3.html

snapd only accepts $6 i.e. SHA-512: https://github.com/snapcore/snapd/blob/master/asserts/user.go#L177